### PR TITLE
duplicate detection for add commands

### PR DESCRIPTION
**Pull request by mandelsoft**
## Description

The add commands now check for duplicate ids for resources, sources and references in specs and report
appropriate problems with the location in the source yaml files.
The same check is done for adding multiple components to avoid overwrite of components from the same action.

The CD validation now reports the identity of the validated component version, if the validation fails.

Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [x] 🐛 Bug Fix
- [x] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [x] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

- Related Issue # (issue)
- Closes # (issue)
- Fixes # (issue)
> Remove if not applicable

## Added tests?

- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help
- [ ] Separate ticket for tests # (issue/pr)

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Added to documentation?

- [ ] 📜 README.md
- [ ] 🙅 no documentation needed

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: cmds/ocm/commands/ocmcmds/common/addhdlrs/comp/elements.go [hunk 1]
```diff
@@ -155,6 +155,10 @@ type ResourceSpec struct {
 
 var _ addhdlrs.ElementSpec = (*ResourceSpec)(nil)
 
+func (r *ResourceSpec) GetRawIdentity() metav1.Identity {
+	return metav1.NewIdentity(r.Name, metav1.SystemIdentityVersion, r.Version)
```
Comment (fabianburth) [inline cmds/ocm/commands/ocmcmds/common/addhdlrs/comp/elements.go:159]: Why is this constant called `SystemIdentityVersion`?

**Review by fabianburth — CHANGES_REQUESTED** [review]

File: cmds/ocm/commands/ocmcmds/common/addhdlrs/utils.go [hunk 2]
```diff
@@ -279,6 +283,39 @@ func Validate(r *ResourceInput, ctx inputs.Context, inputFilePath string) error
 	return allErrs.ToAggregate()
 }
 
+func ValdateElementIdentities(kind string, elems []Element) error {
```
[the remark this answers is not in the document]
Comment (mandelsoft) [reply cmds/ocm/commands/ocmcmds/common/addhdlrs/utils.go]: fixed

File: cmds/ocm/commands/ocmcmds/common/addhdlrs/utils.go [hunk 3]
```diff
@@ -279,6 +283,39 @@ func Validate(r *ResourceInput, ctx inputs.Context, inputFilePath string) error
...
+		dig := id.Digest()
+		if s, ok := ids[string(dig)]; ok {
+			list.Add(fmt.Errorf("duplicate %s identity %s (%s and %s)", kind, id, r.Source(), s))
+		}
+		ids[string(dig)] = r.Source()
+	}
+	return list.Result()
+}
+
+func ValdateElementSpecIdentities(kind string, src string, elems []ElementSpec) error {
```
[the remark this answers is not in the document]
Comment (mandelsoft) [reply cmds/ocm/commands/ocmcmds/common/addhdlrs/utils.go]: added doc to various types and functions

File: cmds/ocm/commands/ocmcmds/common/addhdlrs/comp/elements.go (hunk 1 above)
Comment (mandelsoft) [reply cmds/ocm/commands/ocmcmds/common/addhdlrs/comp/elements.go:159]: SystemIdentityVersion and SystemIdentityName are the two predefined identity properties use to identify elements (resources, sources, referebces) in CDs.

File: cmds/ocm/commands/ocmcmds/common/addhdlrs/comp/components.go [hunk 4]
```diff
@@ -19,6 +19,19 @@ import (
 )
 
 func ProcessComponents(ctx clictx.Context, ictx inputs.Context, repo ocm.Repository, complete ocm.ComponentVersionResolver, thdlr transferhandler.TransferHandler, h *ResourceSpecHandler, elems []addhdlrs.Element) (err error) {
+	list := errors.ErrorList{}
+
+	for _, elem := range elems {
+		if r, ok := elem.Spec().(*ResourceSpec); ok {
+			list.Add(addhdlrs.ValdateElementSpecIdentities("resource", elem.Source().String(), generics.ConvertSliceTo[addhdlrs.ElementSpec](r.Resources)))
+			list.Add(addhdlrs.ValdateElementSpecIdentities("source", elem.Source().String(), generics.ConvertSliceTo[addhdlrs.ElementSpec](r.Sources)))
+			list.Add(addhdlrs.ValdateElementSpecIdentities("reference", elem.Source().String(), generics.ConvertSliceTo[addhdlrs.ElementSpec](r.References)))
```
[the remark this answers is not in the document]
Comment (mandelsoft) [reply cmds/ocm/commands/ocmcmds/common/addhdlrs/comp/components.go]: added docu

File: cmds/ocm/commands/ocmcmds/common/addhdlrs/comp/elements.go (hunk 1 above)
Comment (mandelsoft) [reply cmds/ocm/commands/ocmcmds/common/addhdlrs/comp/elements.go:159]: added docu to their const definitions.

**Pull request merged**